legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BitemporalDeltaPlanner.java

```diff
@@ -436,77 +447,73 @@ private Insert getStageToTemp()
+            selectX1 = Selection.builder().source(stagingDataset()).addAllFields(primaryKeyFieldsAndFromFieldFromStage).condition(dataSplitInRangeCondition).alias(LEFT_DATASET_IN_JOIN_ALIAS).build();
         }
         else
         {
-            selectX1 = Selection.builder().source(stagingDataset()).addAllFields(primaryKeyFieldsAndFromField).alias(LEFT_DATASET_IN_JOIN_ALIAS).build();
+            selectX1 = Selection.builder().source(stagingDataset()).addAllFields(primaryKeyFieldsAndFromFieldFromStage).alias(LEFT_DATASET_IN_JOIN_ALIAS).build();
         }
-        Selection selectY1 = Selection.builder().source(mainDataset()).condition(openRecordCondition).addAllFields(primaryKeyFieldsAndFromField).alias(RIGHT_DATASET_IN_JOIN_ALIAS).build();
+        Selection selectY1 = Selection.builder().source(mainDataset()).condition(openRecordCondition).addAllFields(primaryKeyFieldsAndFromFieldFromMain).alias(RIGHT_DATASET_IN_JOIN_ALIAS).build();
 
         Condition x1AndY1PkMatchCondition = LogicalPlanUtils.getPrimaryKeyMatchCondition(selectX1, selectY1, primaryKeys.toArray(new String[0]));
-        Condition x1FromLessThanY1From = LessThan.of(sourceValidDatetimeFrom.withDatasetRef(selectX1.datasetReference()), sourceValidDatetimeFrom.withDatasetRef(selectY1.datasetReference()));
+        Condition x1FromLessThanY1From = LessThan.of(FieldValue.builder().datasetRef(selectX1.datasetReference()).fieldName(VALID_DATE_TIME_FROM_NAME).build(), FieldValue.builder().datasetRef(selectY1.datasetReference()).fieldName(VALID_DATE_TIME_FROM_NAME).build());
         Condition joinXY1Condition = And.builder().addConditions(x1AndY1PkMatchCondition, x1FromLessThanY1From).build();
         Join joinXY1 = Join.of(selectX1, selectY1, joinXY1Condition, JoinOperation.LEFT_OUTER_JOIN);
 
-        List<Value> selectXY1Fields = new ArrayList<>();
-        selectXY1Fields.addAll(primaryKeyFields.stream().map(field -> field.withDatasetRef(selectX1.datasetReference())).collect(Collectors.toList()));
-        selectXY1Fields.add(sourceValidDatetimeFrom.withDatasetRef(selectX1.datasetReference()));
+        List<Value> selectXY1Fields = primaryKeyFieldsAndFromFieldForSelection.stream().map(field -> field.withDatasetRef(selectX1.datasetReference())).collect(Collectors.toList());
         selectXY1Fields.add(FunctionImpl.builder()
             .functionName(FunctionName.COALESCE)
-            .addValue(FunctionImpl.builder().functionName(FunctionName.MIN).addValue(sourceValidDatetimeFrom.withDatasetRef(selectY1.datasetReference())).build(), LogicalPlanUtils.INFINITE_BATCH_TIME())
+            .addValue(FunctionImpl.builder().functionName(FunctionName.MIN).addValue(FieldValue.builder().datasetRef(selectY1.datasetReference()).fieldName(VALID_DATE_TIME_FROM_NAME).build()).build(), LogicalPlanUtils.INFINITE_BATCH_TIME())
```

Comment:
Same comment as above, define a FieldValue and use it withDatasetRef every where